Per sample unjoined loss

deniskrut opened the pull request:
Summary:
Implements per sample unjoined loss.

In unjoined training data each positive sample has equivalent negative sample. This negative sample needs to be ignored. Special loss function is necessary to achieve this.

This loss function allows mixture of "unjoined" training data and "joined" (the usual) training data in one pipeline. Samples from each can be differenciated using special supervision signal.

Differential Revision: D19602938

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit c9f764cd:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c9f764cd2744360e7e482ad8b707bde57cf522ee).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4361003) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92088769) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92088769))
<details>
<summary>
<code>Jan 28 11:41:32 RuntimeError: test_autograd failed!</code>
</summary>

```
Jan 28 11:41:32 Ran 1046 tests in 286.570s 
Jan 28 11:41:32  
Jan 28 11:41:32 FAILED (errors=1, skipped=13, expected failures=1) 
Jan 28 11:41:32  
Jan 28 11:41:32 Generating XML reports... 
Jan 28 11:41:32 Traceback (most recent call last): 
Jan 28 11:41:32   File "test/run_test.py", line 456, in <module> 
Jan 28 11:41:32     main() 
Jan 28 11:41:32   File "test/run_test.py", line 449, in main 
Jan 28 11:41:32     raise RuntimeError(message) 
Jan 28 11:41:32 RuntimeError: test_autograd failed! 
Jan 28 11:41:32 + cleanup 
Jan 28 11:41:32 + retcode=1 
Jan 28 11:41:32 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Per sample unjoined loss (#32714)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32714

Implements per sample unjoined loss.

In unjoined training data each positive sample has equivalent negative sample. This negative sample needs to be ignored. Special loss function is necessary to achieve this.

This loss function allows mixture of "unjoined" training data and "joined" (the usual) training data in one pipeline. Samples from each can be differenciated using special supervision signal.

Differential Revision: D19602938

fbshipit-source-id: 8f82b4de81885dbffa3780e0f51411f1d5e04438
